staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -399,7 +401,54 @@ func (s *store) List(ctx context.Context, key, resourceVersion string, pred stor
 	if !strings.HasSuffix(key, "/") {
 		key += "/"
 	}
-	getResp, err := s.client.KV.Get(ctx, key, clientv3.WithPrefix())
+	keyPrefix := key
+
+	var fromKey string
+	var fromRV int64
+	if len(pred.From) > 0 {
+		out, err := base64.RawURLEncoding.DecodeString(pred.From)
+		if err != nil {
+			return fmt.Errorf("from key is not valid: %v", err)
+		}
+		from := string(out)
+		cleaned := path.Clean(from)
+		if cleaned != from || cleaned == "." || cleaned == "/" {
+			return fmt.Errorf("from key is not valid")
+		}
+		if strings.HasPrefix(cleaned, "/") {
+			cleaned = cleaned[1:]
+		}
+		if len(cleaned) == 0 {
+			return fmt.Errorf("from key is not valid")
+		}
+		parts := strings.SplitN(cleaned, "/", 2)
+		if len(parts) != 2 {
+			return fmt.Errorf("from is not valid")
+		}
+		fromRV, err = strconv.ParseInt(parts[0], 10, 64)
+		if err != nil {
+			return fmt.Errorf("from resource version is not valid")
+		}
+		fromKey = keyPrefix + parts[1] + "\x00"
+	}
+
+	var options []clientv3.OpOption
+	switch {
+	case len(fromKey) > 0 && pred.Limit > 0:
+		options = []clientv3.OpOption{clientv3.WithRev(fromRV), clientv3.WithRange(key + "\xFF"), clientv3.WithLimit(pred.Limit)}
```

Comment:
@xiang90 if I want to prevent seeing updates newer than `fromRV`, what do I filter by?